pkg/controllers/provisioning/provisioner.go

```diff
@@ -161,32 +161,34 @@ func (p *Provisioner) getPods(ctx context.Context) ([]*v1.Pod, error) {
 func (p *Provisioner) schedule(ctx context.Context, pods []*v1.Pod) ([]*scheduler.Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 
-	// Get instance type options
-	instanceTypes, err := p.cloudProvider.GetInstanceTypes(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("getting instance types, %w", err)
-	}
-
 	// Build node templates
 	var nodeTemplates []*scheduling.NodeTemplate
 	var provisionerList v1alpha5.ProvisionerList
+	instanceTypes := map[string][]cloudprovider.InstanceType{}
 	if err := p.kubeClient.List(ctx, &provisionerList); err != nil {
 		return nil, fmt.Errorf("listing provisioners, %w", err)
 	}
 	for i := range provisionerList.Items {
-		requirements, err := p.cloudProvider.GetRequirements(ctx, provisionerList.Items[i].Spec.Provider)
+		provisioner := &provisionerList.Items[i]
+		requirements, err := p.cloudProvider.GetRequirements(ctx, provisioner.Spec.Provider)
 		if err != nil {
 			return nil, fmt.Errorf("getting provider requirements, %w", err)
 		}
-		nodeTemplates = append(nodeTemplates, scheduling.NewNodeTemplate(&provisionerList.Items[i], requirements))
+		// Get instance type options
+		instanceType, err := p.cloudProvider.GetInstanceTypes(ctx, provisioner.Spec.Provider)
+		if err != nil {
+			return nil, fmt.Errorf("getting instance types, %w", err)
+		}
+		instanceTypes[provisioner.Name] = append(instanceTypes[provisioner.Name], instanceType...)
+		nodeTemplates = append(nodeTemplates, scheduling.NewNodeTemplate(provisioner, requirements))
```

Comment:
Curious if it would make sense to add instanceTypes as a field to the nodeTemplates type.  I think that would prevent having to use a map to store provider specific instanceTypes.

(Similarly, maybe we could add DaemonSet overhead as well)

Reply:
Yeah this sounds reasonable to me. We might want to decouple it from what's passed to the cloudprovider, but it's not the end of the world to expose this information. I'd start by thinking about the fundamental role that NodeTemplate plays. Try explaining to a 5 year old what its role is.

Reply:
IIUC, this will create a circular dependency unless we make other changes.